pkg/services/query/query.go

```diff
@@ -77,27 +85,71 @@ func (s *ServiceImpl) Run(ctx context.Context) error {
 }
 
 // QueryData processes queries and returns query responses. It handles queries to single or mixed datasources, as well as expressions.
-func (s *ServiceImpl) QueryData(ctx context.Context, user *user.SignedInUser, skipCache bool, reqDTO dtos.MetricRequest) (*backend.QueryDataResponse, error) {
+// By default, it will attempt to retrieve a cached response for the given query, unless `skipQueryCache` = true.
+// This is distinct from the `skipDSCache` argument, which only determines if the datasource metadata is cached.
+// Query caching is only implemented in Grafana Enterprise.
+func (s *ServiceImpl) QueryData(ctx context.Context, user *user.SignedInUser, skipDSCache bool, skipQueryCache bool, reqDTO dtos.MetricRequest) (*backend.QueryDataResponse, error) {
+	start := time.Now() // time how long this request takes
+
+	// First look in the query cache if enabled
+	var cr caching.CachedDataResponse
+	if !skipQueryCache {
+		if r := s.cachingService.HandleQueryRequest(ctx, reqDTO); r.Status == caching.StatusCacheHit {
+			caching.MarkCacheStatus(skipQueryCache, r)
+			return r.Response, nil
+		} else {
+			cr = r
+		}
+	}
+
+	// do the actual queries
+	resp, err := s.queryData(ctx, user, skipDSCache, reqDTO)
+
+	// Update the query cache with the result for this metrics request
+	if err == nil {
+		if cr.UpdateCacheFn != nil {
```

Comment:
Made this choice for two reasons:
- Easy to understand by devs
- Avoids having to perform the same logic twice in the enterprise code (once to check cache, another to receive it, this lets us keep stuff in this function's scope)